andi/andi.py

```diff
@@ -1,62 +1,302 @@
+    >>> assert type(c) is C
+    >>> assert c.a is instances[A]
+    >>> assert c.b is instances[B]
+    >>> assert c.b.a is instances[A] # Instance of A is reused (singleton)
+
+    :param func: Function to be inspected.
+    :param is_injectable: A predicate or a container that says if a type
+        is injectable. The planer is responsible to deal
+        with all types that are injectable when traversing
+        the dependency graph. It fails with
+        ``NonProvidableError`` if it is not possible to generate
+        a plan for any of the injectable type found during the inspection.
+    :param externally_provided: A predicate or a dictionary that says if
+        the value for the class/function will be provided externally.
+        The planner won't try to resolve
+        its dependencies, so it acts as a way to stop dependency injection
+        for these classes/functions where we don't want it because they will be
+        provided by other means.
+    :param strict: If the argument ``strict`` is True then this function fails
+        with ``NonProvidableError`` if not all the required arguments
+        for the input class/function can be resolved. When ``strict`` is False,
+        this function provides the plan only for those arguments that could
+        be resolved, so the last task of the plan could be incomplete.
+    :return: A plan
     """
-    if callable(arguments_or_func):
-        arguments = inspect(arguments_or_func)
-    else:
-        arguments = arguments_or_func
+    is_injectable = _ensure_can_provide_func(is_injectable)
+    externally_provided = _ensure_can_provide_func(externally_provided)
+
+    plan_odict = _plan(class_or_func,
+                 is_injectable=is_injectable,
+                 externally_provided=externally_provided,
+                 strict=strict,
+                 dependency_stack=None)
+    return Plan(plan_odict.items())
+
+
+_PlanDict = MutableMapping[Callable, Dict[str, Callable]]
+
+def _plan(class_or_func: Callable, *,
+          is_injectable: Callable[[Callable], bool],
+          externally_provided: Callable[[Callable], bool],
+          strict,
+          dependency_stack=None) -> _PlanDict:
+    dependency_stack = dependency_stack or []
+    plan_seq  = OrderedDict()  # type: _PlanDict
+    type_for_arg = {}
+
+    if externally_provided(class_or_func):
+        plan_seq[class_or_func] = {}
+        return plan_seq
 
-    if isinstance(can_provide, Container):
-        can_provide = can_provide.__contains__
+    if dependency_stack and not is_injectable(class_or_func):
+        raise NonProvidableError(
+            "Type {} cannot be provided".format(class_or_func))
+
+    if class_or_func in dependency_stack:
+        raise CyclicDependencyError(
+            "Cyclic dependency found. Dependency graph: {}".format(
+                " -> ".join(map(str, dependency_stack + [class_or_func]))))
+
+    dependency_stack = dependency_stack + [class_or_func]
```

Comment:
```suggestion
    dependency_stack.append(class_or_func)
```
I don't think it matters here, but why not ensure a temporary list is not created, making it O(N^2)? I'm not sure it is O(N^2) even now, but .append is not worse, and a bit shorter :)

Reply:
Hm, or maybe you don't want dependency_stack to be changed in-place.. sory for noise :)

Reply:
Yes, we need a new list in each iteration.